### PR TITLE
Gitlab CI

### PR DESCRIPTION
**Pull request by SkySkimmer**
Example here https://gitlab.com/SkySkimmer/coq/pipelines/8579411/builds (I changed the cache path and added it to FIND_VCS_CLAUSE after that commit)
It reuses the scripts in `dev/ci`, with a small change to work when Coq is installed with `-prefix install/`, so it should keep up with travis easily.
The jobs in stage `build` make the binaries and stdlib. Each `ci-` job, the test-suite and the documentation reuse the product through build artifacts (and the documentation is published as an artifact). The `warnings` job build with `-warn-error` independently.
This should check as many things as travis, except for fiat-crypto which doesn't have enough memory on gitlab's machines.

Making the doc from an installed Coq is a bit hacky, but I'm not good enough at Makefile to do better.

### CODE REVIEW

**Comment by Zimmi48** [comment]
That's nice and it can't hurt to have this additional file in the repo.
I think you have also looked at Travis build stage. Was it simpler for you to add build stages with GitLab-CI than with Travis?

**Comment by SkySkimmer** [comment]
The issue with Travis build stages is that I have no good way to send data (the compiled Coq) from one job to another. There are basically 2 non-good possibilities:
- using the cache. I experimented with this on a branch at https://github.com/SkySkimmer/coq/tree/travis-stages (first full featured build https://travis-ci.org/SkySkimmer/coq/builds/234566593, recent rebase https://travis-ci.org/SkySkimmer/coq/builds/236676732).
As you can see the build targets aren't set by environment variables any more, and so the travis page is unreadable. This is because jobs with different environments get isolated caches.
There is also an issue with jobs possibly interfering with each other, eg I noticed that ssreflect was getting installed in the cache. This means that for instance if we have
```
X --- Y means Y uses products of X

       --B--
      /     \
     /       \
A ---         ---D
     \       
      \     
       --C
```
where B and C add something from the cache, if C happens after B it will overwrite B's additions and D will not get them (depending on how the race goes). So while using the cache works OK for linear additions then parallel noops it wouldn't work to expand the mechanism to reuse eg math-classes when building formal-topology (this isn't implemented for gitlab in this PR either, but there I think it would be feasible).

- sending files to some 3rd party service like amazon. That would have a money cost, and beyond that it doesn't work with developer repositories and PRs: travis would let us have a secret access key for the server, but it wouldn't be accessible to pull requests and forks so it would be of limited benefit. See https://docs.travis-ci.com/user/pull-requests#Pull-Requests-and-Security-Restrictions


So in summary I don't think we can get much benefit from Travis build stages unless they get a system for build artifacts.

**Comment by Zimmi48** [comment]
Thanks!
Since build stages are pretty new and they say "We'd love your feedback." I think it would be worth sending them what you wrote.

I think your objection to the idea of relying on deployment on a 3rd party service is fundamental indeed. Automatic deployment is a very useful thing to have, but you're showing that it has necessarily to be distinguished from automatic testing.

At least for now, your PR brings a feature that won't be really useful here on GitHub but anyone can benefit from it by having a fork in GitLab and running tests here (before submitting a PR), because the results will be faster. Pushing to a fork on GitLab for the automatic testing has the additional advantage that it won't show up in the right panel on Gitter.

**Comment by SkySkimmer** [comment]
I think build stages are a logical constraint, with the basic example being "stage 1: test, stage 2: publish, but only if tests succeeded". However Coq is more set up for practical constraints like make dependencies, where the result of one job is necessary to run the next.
The current system of Travis doesn't really work for this, and the one of gitlab forces us to do manual scheduling through build stages. Gitlab has some issues about this (https://gitlab.com/gitlab-org/gitlab-ce/issues/29347 https://gitlab.com/gitlab-org/gitlab-ce/issues/28592) but I don't think there's anything imminent. Travis has https://github.com/travis-ci/beta-features/issues/11#issuecomment-304086106
>Jobs do not share any storage at the moment. In order to share build artifacts (for example a binary compiled in an earlier stage) you can do so using our artifacts feature (see https://docs.travis-ci.com/user/uploading-artifacts/), or manage the process manually, e.g. https://docs.travis-ci.com/user/build-stages/share-files-s3/. We do intend to improve on this in the future.

**Comment by SkySkimmer** [comment]
>At least for now, your PR brings a feature that won't be really useful here on GitHub but anyone can benefit from it by having a fork in GitLab and running tests here (before submitting a PR), because the results will be faster. Pushing to a fork on GitLab for the automatic testing has the additional advantage that it won't show up in the right panel on Gitter.

Is that an advantage? 😋 
I'm not sure how timings compare. Gitlab machines seem to have somewhat inconsistent timings, generally slower than travis (it might just be the `sudo: required` issue I don't know). They do offer unrestricted parallel jobs vs Travis limit of 5, so I think my avoiding duplicate work is actually losing time spinning up containers 😆 OTOH when I started experimenting they had some issue with their CI machine provider (https://gitlab.com/gitlab-com/infrastructure/issues/1838) so my impression of slowness may be caused by my first impression of waiting a while for available machines.

**Comment by ejgallego** [comment]
I think that using S3 may be fine thou.

**Comment by Zimmi48** [comment]
@ejgallego 
> I think that using S3 may be fine thou.

Can you elaborate how this would compose with testing PRs and forks?

@SkySkimmer 
> I'm not sure how timings compare. Gitlab machines seem to have somewhat inconsistent timings, generally slower than travis...

Oh well, so it's really more a simple experiment with no real immediate practical purpose, isn't it?

**Comment by ejgallego** [comment]
> Can you elaborate how this would compose with testing PRs and forks?

Certainly not very optimally, but it could be set up so all of them use the same S3.

**Comment by SkySkimmer** [comment]
Looking at it I was definitely wrong about it being slower than travis: these 2 builds are for this PR, pushed at the same time:
- Gitlab: https://gitlab.com/SkySkimmer/coq/pipelines/8580261 says it took 57 minutes 41 seconds
- Travis: https://travis-ci.org/SkySkimmer/coq/builds/236672566 running for 1h 30min with 4 jobs not yet started, probably ~30min left.

Past travis trunk took 2h or more, my experiment with travis stages took 1h 40min.

I'm not sure how time starting up machines is counted but it should impact both and the simultaneous push comparison doesn't rely on the time measurement.

**Comment by JasonGross** [comment]
> except for fiat-crypto which doesn't have enough memory on gitlab's
machines.

This is a bug with par, right?  At the very least, it should be fixed by
having par fall-back to all: when fork gives memory errors, no?

**Comment by maximedenes** [comment]
@SkySkimmer Can you say a bit about the motivation for this?

The only thing that worries me a bit is that I wouldn't like to have to make each change to the Travis infrastructure a second time. So who's going to maintain `.gitlab-ci.yml` ?

File: Makefile.ci [hunk 1]
```diff
@@ -8,3 +8,8 @@ CI_TARGETS=ci-all ci-hott ci-math-comp ci-compcert ci-sf ci-cpdt \
 # Generic rule, we use make to easy travis integraton with mixed rules
 $(CI_TARGETS): ci-%:
 	./dev/ci/ci-$*.sh
+
+# For emacs:
+# Local Variables:
+# mode: makefile
+# End:
```
Comment (Zimmi48) [inline Makefile.ci]: Is this really something you wanted to leave here?

File: test-suite/Makefile [hunk 2]
```diff
@@ -21,16 +21,18 @@
 # The "run" target runs all tests that have not been run yet. To force
 # all tests to be run, use the "clean" target.
 
+# TODO https://blog.flameeyes.eu/2008/11/for-a-parallel-world-case-study-n-4-jobserver-unavailable/
+
```
Comment (Zimmi48) [inline test-suite/Makefile]: Same question here.

File: Makefile.ci (hunk 1 above)
Comment (SkySkimmer) [inline Makefile.ci]: Yes, but I'm reconsidering it since it's the only change for this file.

File: test-suite/Makefile (hunk 2 above)
Comment (SkySkimmer) [inline test-suite/Makefile]: No.

**Comment by SkySkimmer** [comment]
@maximedenes 
This makes Coq less dependent on specific (and non-free) infrastructure, tests `make install`, and provides a way to run CI in (guessing from a small amount of runs) 30% to 50% less time to anyone who clones their repository to gitlab.

The changes that would have to be duplicated would be the ones to the travis.yml itself. There may also be changes which break tests with installed Coq but not with `-boot` Coq.
From history changes to travis.yml are either new targets or things specific to travis. New targets are easy to duplicate.
I'm not sure how often things will break. I have set my gitlab to mirror my github so I should be alerted eventually if it happens.

I should add something to README.ci.

**Comment by Zimmi48** [comment]
> I should add something to README.ci.

Yes, please do that. And add advice on how to use GitLab to take advantage of it if you have any. E.g. is mirroring a good idea in your opinion?

**Comment by SkySkimmer** [comment]
I don't know enough about mirroring to tell, I just set it up the other day.

File: README.ci [hunk 3]
```diff
@@ -75,3 +76,29 @@ The `.overlay` file will contain a set of variables that will be used
...
+Travis rebuilds all of Coq's executables and stdlib for each job. Coq
+is built with `./configure -local`, then used for the job's test.
+
+Gitlab specific information
+===========================
+
+Gitlab is set up to use the "build artifact" feature to avoid
+rebuilding Coq. In one job, Coq is built with `./configure -prefix
+install` and `make install` is run, then the `install` directory
+persists to and is used by the next jobs.
```
Comment (Zimmi48) [inline README.ci]: persists to? I'm not sure the "to" is necessary here or something is missing. (Although I'm not a native English speaker so if you tell me this is fine, I cannot argue.)

File: README.ci [hunk 4]
```diff
@@ -6,15 +6,16 @@ Introduction
 
 The Coq Travis CI infrastructure is meant to provide lightweight
 automatics testing of pull requests.
+If you are on Gitlab, their integrated CI is also set up.
```
Comment (Zimmi48) [inline README.ci]: The official way of writing Gitlab is GitLab.

File: README.ci [hunk 5]
```diff
@@ -75,3 +76,29 @@ The `.overlay` file will contain a set of variables that will be used
 to do the corresponding `opam pin` or to overload the corresponding
 git repositories, etc...
 
+Since pull requests only happen on Github there is no need to test the
```
Comment (Zimmi48) [inline README.ci]: The official way of writing Github is GitHub.

**Comment by maximedenes** [comment]
@SkySkimmer @ejgallego Can I merge this?

**Comment by SkySkimmer** [comment]
Yes as far as I'm concerned.

**Comment by ejgallego** [comment]
LGTM
